controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -238,41 +239,48 @@ public String getScopeName() {
     }
 
     private CompletableFuture<List<Segment>> getSuccessorsForSegment(final int number) {
-        val segmentFuture = getSegment(number);
-        val indexTableFuture = getIndexTable();
-        val historyTableFuture = getHistoryTable();
-        CompletableFuture<Void> all = CompletableFuture.allOf(segmentFuture, indexTableFuture, historyTableFuture);
-
-        return all.thenCompose(x -> {
-            final Segment segment = segmentFuture.getNow(null);
-            List<Integer> candidates = TableHelper.findSegmentSuccessorCandidates(segment,
-                    indexTableFuture.getNow(null).getData(),
-                    historyTableFuture.getNow(null).getData());
-            return findOverlapping(segment, candidates);
-        });
+        return getHistoryTable()
+                .thenApply(historyTable -> {
+                    CompletableFuture<Segment> segmentFuture = getSegment(number);
+                    CompletableFuture<Data<T>> indexTableFuture = getIndexTable();
+                    return new ImmutableTriple<>(historyTable, segmentFuture, indexTableFuture);
```

Comment:
You can change this `thenApply` to `thenCompose` and return `CompletableFuture.allOf` instead of `ImmutableTriple`. The behavior is identical but your code will be much cleaner and you'll have fewer callbacks to worry about.

Reply:
Oh I see what you did. You actually need the result of all of them. Then You can use `FutureHelpers.allOfWithResults` which will return an array (or list) in the same order in which you passed them to the function.

Reply:
actually not all of them are future.. first one is data and other two are futures. So using FutureHelpers.allOfWithResults would still need to be paired up with the data..